woocommerce/woocommerce
# Add webhook delivery retries (exponential backoff)

### All Submissions:






### Changes proposed in this Pull Request:

Add retry/reschedule logic for Webhooks.

### How to test the changes in this Pull Request:

1. Update the woo tables. _I'm not sure of the best way, but I deactivated the plugin, dropped the woo tables and then reactivated the plugin._
1. Start a basic web server that only returns 200 responses: `docker run -it -p 8000:80 --rm bnef/echo-server`
1. Map an external URL to your server (e.g. `ngrok http 8000`). For some reason both `localhost` and `host.docker.internal` didn't work for me.
1. In a browser create a webhook (WooCommerce → Settings → Advanced → Webhooks) by adding a name, topic, and delivery url. I used a `coupon.updated` topic, and the delivery url should be your ngrok address (above). Make sure the status is Active.
1. Enable retries for this webhook (change the webhook id): `wp eval '$hook = new WC_Webhook(1); $hook->set_retry_enabled( true ); $hook->save();'`
1. Create a new coupon
1. Open a new window in the Scheduled Actions admin (Tools → Scheduled Actions).
1. Find the scheduled action for your coupon (looks similar to this): ![](https://user-images.githubusercontent.com/6354169/87343455-bcb77400-c501-11ea-8247-32135461f58d.png)
1. When you mouse over the row you'll see the option to "Run" the webhook, run it and verify in your terminal window (step 2) that you receive the webhook.
1. Stop the terminal process and restart it to return 500 responses: `docker run -it -p 8000:80 --rm bnef/echo-server 500`.
1. Update the coupon, return to the "Scheduled Actions" page and run the new scheduled action -- this will fail and refreshing the page will show a new pending action scheduled in the future. If you try running this one, the next one scheduled will schedule farther out in the future (this repeats up to 5 times).



### Other information:




### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

* Added webhook retry capability (with exponential backoff).

cc @ObliviousHarmony
